Introduce initial support for NXP S32Z27 SoC

manuargue opened the pull request:
This PR introduces initial support for NXP S32Z27 SoC (based on Cortex-R52), including:

- SoC and board support
- Shim drivers for UART (poll and interrupt-based APIs), GPIO (basic APIs), and Pinctrl
- Lauterbach TRACE32 West runner.

Currently is only possible to use Lauterbach TRACE32 West runner to load and execute a Zephyr application from internal SRAM. WIP to be addressed in future PR's: run from Flash, implement all GPIO APIs.

The updates for HAL NXP are in https://github.com/zephyrproject-rtos/hal_nxp/pull/182

manuargue-nxp committed:
scripts: west: introduce Lauterbach TRACE32 runner

This patch introduces a West runner for flashing and debugging with
Lauterbach TRACE32 debuggers. The runner consists of a wrapper around
TRACE32 software, and allows a Zephyr board to execute a custom start-up
script (Practice Script) for the different commands supported, including
the ability to pass extra arguments from CMake. Is up to the board using
this runner to define the actions performed on each command.

The `debug` command launches TRACE32 GUI to allow debug the Zephyr
application, while the `flash` command hides the GUI and executes the
start-up script in a background process.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
dts: arm: add ARMv8-R skeleton

Introduce base dtsi for ARMv8-R AArch32 based SoCs.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
soc: arm: introduce NXP S32Z/E support

This patch introduces support for NXP S32 devices, specifically for
S32Z27 from S32Z/E family.

NXP S32Z27 processors are composed of two Real-Time Units (RTU)
containing each four ARM Cortex-R52 cores with flexible split/lock
configuration, and dedicated internal SRAM.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: pinctrl: add NXP S32 SIUL2 driver

Introduce Pinctrl driver for SIUL2 module present on NXP S32 devices,
which provides control over all pins, such as function selection and
electrical characteristics that appear on external chip pins.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: arm: introduce s32z270dc2_r52 support

This patch introduces the support for NXP X-S32Z27X-DC(2) board.
Each S32Z27 Real-Time Unit (RTU) is defined as a separate board to
allow the user to run Zephyr on a specific RTU.

Currently is only possible to use Lauterbach TRACE32 west runner to load
and execute a Zephyr application from internal SRAM.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
soc: arm: add pinctrl support for NXP S32Z/E SoC's

Introduce support for Pinctrl driver on NXP S32Z/E SoC's.
The NXP S32 pin controller is a singleton node responsible for
controlling the pin function selection and pin properties, based on the
pin node group approach. The pinmux configuration is encoded in a
32-bit value.

Each S32 SoC implementing Pinctrl must create a `pinctrl_soc.h` header
which define SoC-specific macros to initialize the pinctrl structure.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: arm: s32z270dc2_r52: enable pinctrl driver

Enable pinctrl driver, configuration for pinctrl node to be added in
future patches.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

Dat-NguyenDuy committed:
drivers: serial: support NXP S32 LINFlexD UART

This patch introduces support for NXP S32 LINFlexD peripheral operating
in UART mode. Polling and interrupt-based serial API's are supported.

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>
Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: arm: s32z270dc2_r52: enable serial driver

Enable serial driver and Zephyr console on this board.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: gpio: add support for NXP S32 devices

Implement GPIO driver minimal API's for NXP S32 devices, based on SIUL2
peripheral. SIUL2 allows to control the pins electrical characteristics
such as internal pull resistors, pin direction and more.

GPIO driver API's for interrupts will be implemented in a future patch.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: arm: s32z270dc2_r52: enable GPIO driver

Enable GPIO driver on s32z270dc2_r52 boards. S32Z27 pads are grouped
into GPIO ports A to N.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
tests: gpio: add overlays for s32z270dc2_r52 boards

Enable testing of GPIO driver minimal API's on s32z270dc2_r52 boards.
These boards do not have built-in user LED, therefore a GPIO pin is used
instead.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: s32z270dc2_r52: add docs

Document supported features and basic how-to for s32z270dc2_r52 boards.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
codeowners: add owner for NXP S32 platforms

Add ownership to NXP S32 platform related files.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

galak commented on drivers/gpio/Kconfig.s32:
```diff
@@ -0,0 +1,11 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_NXP_S32_GPIO := nxp,s32-gpio
```
`DT_COMPAT_NXP_S32_GPIO` is auto-gen'd now, so its not needed.

galak commented on drivers/gpio/Kconfig.s32:
```diff
@@ -0,0 +1,11 @@
+
+config GPIO_S32
+	bool "NXP S32 GPIO driver"
+	default y if $(dt_compat_enabled,$(DT_COMPAT_NXP_S32_GPIO))
```
Change this to:

```
default y
depends on DT_HAS_NXP_S32_GPIO_ENABLED`
```

galak commented on drivers/serial/Kconfig.s32:
```diff
@@ -0,0 +1,29 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_NXP_S32_LINFLEXD := nxp,s32-linflexd
```
See comment in gpio/Kconfig about changes needed, same comment here.

galak commented on modules/Kconfig.s32:
```diff
@@ -0,0 +1,29 @@
+# NXP S32 MCUs family
```
Why is this not part of the modules/Kconfig.mcux

manuargue replied:
The NXP S32 family is not covered by MCUXpresso SDK

galak replied:
But you are using the same HAL repo for it.

galak replied:
> While this is all NXP, we have different business units and product lines. For the moment, the internal guidance has been to separate this so I'm okay with a Kconfig.s32.
> 
> We can consider merging these into a single Kconfig.nxp in a separate PR from this PR.

I don't know if we have examples of how best to handle this pattern - but feels like we should have a 1:1 correspondence with the 'module'.  So if S32 lives in the existing hal_nxp module, than it should be treated as one thing.  @carlescufi do we have cases of different module/Kconfigs that map to one module repo (at this level of the hierarchy)?

gmarull replied:
maybe create `modules/hal_nxp` and multiple `Kconfig.*` inside include by a main `Kconfig`?

galak replied:
As there is a pattern with Kconfig.imx this is ok for now.  However, I agree with @gmarull that all that should get put under a `hal_nxp` dir.  Otherwise its very difficult to know which Kconfig files are associated with which modules.

manuargue replied:
Ok, we will consider this suggestion and address it in a separate PR then.

galak reviewed (changes requested):
Kconfig & dts updates are needed.

mbolivar-nordic commented on boards/common/t32.board.cmake:
```diff
@@ -0,0 +1,6 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+board_set_flasher_ifnset(t32)
```
Please pick a more descriptive/googleable name, such as `trace32`.

manuargue replied:
`trace32` seems more reasonable, updated

mbolivar-nordic commented on scripts/west_commands/runners/t32.py:
```diff
@@ -0,0 +1,143 @@
+    def do_create(cls, cfg: RunnerConfig, args: argparse.Namespace) -> 'TRACE32BinaryRunner':
+        build_conf = BuildConfiguration(cfg.build_dir)
+        if args.arch == 'auto':
+            arch = build_conf.getboolean('CONFIG_ARCH').replace('"', '')
```
getboolean?

manuargue replied:
Nice catch. It was working without problems due to the implementation of `.getboolean`, changed to `.get`.

mbolivar-nordic commented on doc/develop/flash_debug/host-tools.rst:
```diff
@@ -255,6 +255,73 @@ These debug host tools are compatible with the following debug probes:
+Flashing and Debugging
+----------------------
+
+Set the environment variable :envvar:`T32_DIR` pointing to the TRACE32 system
```
Please add a `:ref:` to env_vars.rst instead of restating parts of that page. For instance:

```
Set the :ref:`environment variable <env_vars>` :envvar:`T32_DIR` to ...
```

manuargue replied:
do you suggest to remove the example of setting the envvar and just keep the text above instead?

mbolivar-nordic replied:
>  Do you suggest to also remove the example of setting the envvar and just keep the text above instead?

Yes, please.

manuargue replied:
removed

mbolivar-nordic reviewed (changes requested):
Nice work on the runner. Just a few requests.

povergoing commented on tests/drivers/gpio/gpio_api_1pin/boards/s32z270dc2_rtu1_r52.overlay:
```diff
@@ -0,0 +1,26 @@
+	gpio-led {
+		compatible = "gpio-leds";
+		led_A3: led_A3 {
+			label = "led_A3";
```
ditto

povergoing commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -0,0 +1,407 @@
+			reg = <0>;
+		};
+
+		cpu@1 {
```
Just out of curiosity, do you have booted all 8 cores but there is no SMP supported for v8R32?

manuargue replied:
Not all the cores are boot, only the core on which Zephyr is running single-core. There's the possibility to chose the target core providing different arguments to the west runner as described in https://github.com/zephyrproject-rtos/zephyr/pull/49161/commits/9a7af8f9818f0101b0cbc496017bcf266adccebf#diff-b226f398a3c6bd057366b1fa27e22cac9beac761d9a16e4cab0e36bd9a42e2b7R191-R201

povergoing replied:
Thanks for infomation

povergoing commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -0,0 +1,407 @@
+			status = "okay";
+		};
+
+		sram0: memory@31780000 {
```
Just a quick question, no flash here? Or using sram to simulate flash?

manuargue replied:
There's no embedded flash on this part, we will run from external flash later on (QSPI)

povergoing replied:
Thanks

povergoing commented on soc/arm/nxp_s32/s32ze/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,34 @@
+	default y
+
+config MAX_DOMAIN_PARTITIONS
+	default 24 if USERSPACE
```
Can I ask how many MPU regions your soc have? Do you enable `CONFIG_MPU_GAP_FILLING`? If it is enabled and suppose the regions the soc supported is 32, I guess the `MAX_DOMAIN_PARTITIONS` should be small than 24. Because when you insert a region into the middle of the existing one, you possibly get 2 more regions? That is 24 partitions can cause 40+ regions with `CONFIG_MPU_GAP_FILLING` enabled. It's just a concern, I am okay with it if it turns out to be not an issue :)

manuargue replied:
Hmm thanks for the heads up, I needed to revisit this part. The MPU has 20 regions for EL1 and `CONFIG_MPU_GAP_FILLING` is enabled. Leaving `MAX_DOMAIN_PARTITIONS` to the default of 16 seems a better option.

povergoing reviewed (changes requested):
LGTM overall. Only some suggestions about uart and some simple questions. Besides, I request a change because I saw the dtsi `label`s are still there?

manuargue commented on drivers/serial/uart_s32_linflexd.c:
```diff
@@ -0,0 +1,531 @@
+			CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
+			&uart_s32_driver_api);
+
+#if DT_NODE_HAS_STATUS(UART_S32_NODE(0), okay)
```
The device driver needs to keep track of the hw peripheral instance number as required by the HAL. Using the `DT_INST` approach is indeed less verbose and more compact, but it would require to store the instance number in DT. I'm ok with either way, as long as we stick to it with the rest of the drivers to come.

galak replied:
What you did is the correct way to handle it.  Its unfortunate that the HAL is based on a index.

povergoing replied:
Thanks galak for clarification

manuargue commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -0,0 +1,409 @@
+				<GIC_PPI 11 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
+				<GIC_PPI 10 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+		interrupt-parent = <&gic>;
+		label = "armv8_timer";
```
removed all `labels` from dtsi and bindings

manuargue commented on tests/drivers/gpio/gpio_api_1pin/boards/s32z270dc2_rtu0_r52.overlay:
```diff
@@ -0,0 +1,26 @@
+	gpio-led {
+		compatible = "gpio-leds";
+		led_A3: led_A3 {
+			label = "led_A3";
```
I removed this all all other `labels` in dtsi & bindings. `label` is not needed for this test.

galak replied:
@povergoing label usage for gpio-leds, gpio-keys, pwm-leds, and fixed-partitions are still ok.

ambroise-arm commented:
Hello. I have tried to run the "hello_world" sample with the code from the PR and the associated code for hal_nxp. It compiles fine but while running the app I get the following behavior.
The program goes up to `svc #_SVC_CALL_CONTEXT_SWITCH` in swap_helper.S:916, which correctly leads to `ldr pc, =z_arm_svc` in vector_table.S:22, but instead of jumping to z_arm_svc it goes to `ldr pc, =z_arm_undef_instruction` in vector_table.S:21. And then instead of going to z_arm_undef_instruction it keeps executing that same line forever.
I am not using a Lauterbach debugger, but instead an NXP S32 Debug Probe with initialization scripts from S32 Design Studio. The PR states that only Lauterbach is usable, but I wanted to confirm if what I observed was expected.

ambroise-arm commented:
I couldn't get the modification to work in my startup script, but I could modify `SCTLR.TE` at runtime with the debugger. That way the hello_world sample works for me. That's all I wanted to validate, thanks @manuargue!

povergoing reviewed (dismissed):
LGTM

manuargue commented:
I have rebased to fix the conflicts and updated the HAL on https://github.com/zephyrproject-rtos/hal_nxp/pull/182.
Would be great if this pr can make it before the code freeze next week :)

dleach02 commented:
@galak, @mbolivar-nordic, can you two revisit your comments to determine if they have been addressed

galak commented on modules/Kconfig:
```diff
@@ -28,6 +28,7 @@ source "modules/Kconfig.mcux"
 source "modules/Kconfig.microchip"
 source "modules/Kconfig.nuvoton"
 source "modules/Kconfig.open-amp"
+source "modules/Kconfig.s32"
```
rename to `Kconfig.nxp_s32`

manuargue replied:
I have understood we agreed to address this in a separate PR and leave it as it is for now, please see https://github.com/zephyrproject-rtos/zephyr/pull/49161#discussion_r948042301

dleach02 replied:
@galak, After the 3.2 release, I plan on merging the various "NXP" Kconfig modules into one Kconfig.nxp to address this.

galak commented on drivers/serial/Kconfig.s32:
```diff
@@ -0,0 +1,28 @@
+	bool "LINFlexD UART driver for NXP S32 family processors"
+	default y
+	depends on DT_HAS_NXP_S32_LINFLEXD_ENABLED
+	depends on HAS_S32_LINFLEXD
```
remove `depends on HAS_S32_LINFLEXD`

manuargue replied:
removed

galak commented on drivers/pinctrl/Kconfig.s32:
```diff
@@ -0,0 +1,9 @@
+
+config PINCTRL_S32
+	bool "Pin controller driver for NXP S32 processors"
+	depends on SOC_FAMILY_S32
```
have this depend on the devicetree Kconfig symbol

galak commented on drivers/gpio/Kconfig.s32:
```diff
@@ -0,0 +1,10 @@
+	bool "NXP S32 GPIO driver"
+	default y
+	depends on DT_HAS_NXP_S32_GPIO_ENABLED
+	depends on HAS_S32_GPIO
```
remove `depends on HAS_S32_GPIO`

manuargue replied:
removed

mbolivar-nordic reviewed (dismissed):
approval is for the runner only

gmarull commented on boards/arm/s32z270dc2_r52/board.c:
```diff
@@ -0,0 +1,29 @@
+
+#include <zephyr/init.h>
+
+#include <OsIf.h>
+
+#ifdef CONFIG_INIT_CLOCK_AT_BOOT_TIME
+#include <Clock_Ip.h>
+#include <Clock_Ip_Cfg.h>
+#endif
+
+static int s32_board_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	OsIf_Init(NULL);
+
+#ifdef CONFIG_INIT_CLOCK_AT_BOOT_TIME
+	Clock_Ip_Init(Clock_Ip_aClockConfig);
+#endif
+
+	return 0;
+}
```
why do we need board specific clock config? Can't this be handled by SoC code?

manuargue replied:
You are right. Moved to SoC initialization code.

gmarull commented on boards/arm/s32z270dc2_r52/s32z270dc2_rtu0_r52.yaml:
```diff
@@ -0,0 +1,11 @@
+type: mcu
+arch: arm
+ram: 1024
+flash: 1024 # to allow Twister to run tests from SRAM
```
?

manuargue replied:
I've removed this line. It came from an older of version of Twister we were using internally, but there is not need for it anymore (it was originally to bypass the filter of minimum flash required of some tests since we are running entirely from SRAM).

gmarull commented on drivers/pinctrl/Kconfig.s32:
```diff
@@ -0,0 +1,9 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config PINCTRL_S32
+	bool "Pin controller driver for NXP S32 processors"
+	depends on SOC_FAMILY_S32
+	default y
+	help
```
please use DT_HAS_* helpers here

gmarull commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -98,5 +98,109 @@
+			compatible = "nxp,s32-linflexd";
+			reg = <0x40180000 0x1000>;
+			interrupts = <GIC_SPI 213 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+			label = "UART_1";
```
please remove labels, they have been deprecated

gmarull commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -103,103 +103,90 @@
 			compatible = "nxp,s32-linflexd";
 			reg = <0x40170000 0x1000>;
 			interrupts = <GIC_SPI 212 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
-			label = "UART_0";
```
please amend initial commit

manuargue replied:
amended the wrong commit before :) thanks

gmarull commented on tests/drivers/gpio/gpio_api_1pin/boards/s32z270dc2_rtu0_r52.overlay:
```diff
@@ -0,0 +1,25 @@
+
+	gpio-led {
+		compatible = "gpio-leds";
+		led_A3: led_A3 {
```
nit: `led_a3`

gmarull commented on boards/arm/s32z270dc2_r52/doc/index.rst:
```diff
@@ -0,0 +1,248 @@
+| LINFlexD  | on-chip    | serial                              |
++-----------+------------+-------------------------------------+
+
+The default configuration can be found in the board's defconfig files:
+
+- :zephyr_file:`boards/arm/s32z270dc2_r52/s32z270dc2_rtu0_r52_defconfig`
+- :zephyr_file:`boards/arm/s32z270dc2_r52/s32z270dc2_rtu1_r52_defconfig`.
```
this is redundant, can be deleted.

gmarull commented on boards/arm/s32z270dc2_r52/doc/index.rst:
```diff
@@ -0,0 +1,248 @@
+| PAD_170 - PAD_173 | PO10 - PO13 |
++-------------------+-------------+
+
+The default pinmux configuration can be found in the board's devicetree files:
+
+- :zephyr_file:`boards/arm/s32z270dc2_r52/s32z270dc2_r52-pinctrl.dtsi`
+- :zephyr_file:`boards/arm/s32z270dc2_r52/s32z270dc2_rtu0_r52.dtsi`
+- :zephyr_file:`boards/arm/s32z270dc2_r52/s32z270dc2_rtu1_r52.dtsi`.
```
ditto

gmarull reviewed (changes requested):
some minor comments

manuargue commented:
@gmarull I've addressed your comments. Please resolve them if you consider them done.

cfriedt reviewed (dismissed):
Looks fine for GPIO

gmarull reviewed (dismissed):
please merge hal PR and update west.yaml

galak commented on drivers/serial/uart_s32_linflexd.c:
```diff
@@ -0,0 +1,542 @@
+	data->hw_cfg.TransferType = LINFLEXD_UART_IP_USING_INTERRUPTS;
+	data->hw_cfg.Callback = s32_uart_callback;
+	data->hw_cfg.CallbackParam = NULL;
+	data->hw_cfg.StateStruct = &Linflexd_Uart_Ip_apStateStructure[state_idx++];
```
is state_idx correct?  what happens if I have uart1 and uart5 enabled?

manuargue replied:
it's correct, `Linflexd_Uart_Ip_apStateStructure` length is determined based on the number of instances enabled, and not the total number of instances available. Pls see https://github.com/zephyrproject-rtos/hal_nxp/pull/182/commits/eb424444215a27cbf20053595068bd25d54991db
So if `uart1` and `uart5` are enabled, the length of the array is 2, `uart1` will use index 0 and `uart2` index 1.

galak reviewed (dismissed):
uart: need to cleanup how the "index" of Linflexd_Uart_Ip_apStateStructure is determined.

galak commented:
Please make sure to get @dcpleung for the UART driver.

manuargue commented:
@galak force pushed to rebuild with changes in hal_nxp as requested

dleach02 commented:
@manuargue https://github.com/zephyrproject-rtos/hal_nxp/pull/182 HAL PR has been merged. Please update this PR to reflect the new SHA.

dleach02 commented:
The build failures are due to #51144. I have posted #51145 to fix this.

dleach02 commented:
@manuargue please rebase your PR and push your branch

manuargue commented:
done @dleach02, unrelated failed tests should pass now
